### PR TITLE
Fix MyPy issues in Core airflow packages

### PR DESCRIPTION
**Pull request by potiuk**
Part of #19891

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/example_dags/example_task_group_decorator.py [hunk 1]
```diff
@@ -57,8 +57,8 @@ def task_end():
 
 
 # Creating TaskGroups
-@task_group
-def task_group_function(value):
+@task_group()
```
Comment (potiuk) [inline airflow/example_dags/example_task_group_decorator.py]: Not sure if our task_group decorator also suppots () syntax ?

Comment (uranusjr) [reply airflow/example_dags/example_task_group_decorator.py]: It does

File: airflow/example_dags/example_task_group_decorator.py [hunk 2]
```diff
@@ -33,19 +33,19 @@ def task_start():
 
 
 @task
-def task_1(value):
+def task_1(value: int):
```
Comment (uranusjr) [inline airflow/example_dags/example_task_group_decorator.py]:
```suggestion
def task_1(value: int) -> str:
```

Probably worthwhile to add annotations for return values as well. (Same applies everywhere.)

Comment (potiuk) [reply airflow/example_dags/example_task_group_decorator.py]: Good point

Comment (potiuk) [reply airflow/example_dags/example_task_group_decorator.py]: Done

File: airflow/jobs/scheduler_job.py [hunk 3]
```diff
@@ -1140,8 +1140,8 @@ def _emit_pool_metrics(self, session: Session = None) -> None:
         pools = models.Pool.slots_stats(session=session)
         for pool_name, slot_stats in pools.items():
             Stats.gauge(f'pool.open_slots.{pool_name}', slot_stats["open"])
-            Stats.gauge(f'pool.queued_slots.{pool_name}', slot_stats[TaskInstanceState.QUEUED])
-            Stats.gauge(f'pool.running_slots.{pool_name}', slot_stats[TaskInstanceState.RUNNING])
+            Stats.gauge(f'pool.queued_slots.{pool_name}', slot_stats["queued"])
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job.py]: Maybe `TaskInstanceState.QUEUED.value` is better? (Same for below)

File: airflow/macros/__init__.py [hunk 4]
```diff
@@ -66,7 +67,7 @@ def ds_format(ds: str, input_format: str, output_format: str) -> str:
     return datetime.strptime(str(ds), input_format).strftime(output_format)
 
 
-def datetime_diff_for_humans(dt: Any, since: Optional[datetime] = None) -> str:
+def datetime_diff_for_humans(dt: Any, since: Optional[Union[datetime, DateTime]] = None) -> str:
```
Comment (uranusjr) [inline airflow/macros/__init__.py]: This should not be needed because `Union[datetime, DateTime]` is just `datetime`. This entire module should only need one change; see #20132.

File: airflow/operators/python.py [hunk 5]
```diff
@@ -173,7 +173,9 @@ def __init__(
         self.show_return_value_in_logs = show_return_value_in_logs
 
     def execute(self, context: Context) -> Any:
-        context.update(self.op_kwargs)
+        # The casting below is not `perfect`, but we "pretend" that context is TypedDict via Context.pyi
+        # So no harm done and it will work as expected.
+        context.update(cast(Context, self.op_kwargs))
         context['templates_dict'] = self.templates_dict
```
Comment (uranusjr) [inline airflow/operators/python.py]: Another possibility is

```suggestion
        context_additions = {"templates_dict": self.templates_dict}
        if self.op_kwargs:
            context_additions.update(self.op_kwargs)
            context = Context(context, **context_additions)
```

It’s probably a good idea to add a function in `airflow.utils.context` for this. Maybe call it `context_merge`.

File: airflow/utils/context.pyi [hunk 6]
```diff
@@ -79,6 +79,7 @@ class Context(TypedDict, total=False):
     task_instance: TaskInstance
     task_instance_key_str: str
     test_mode: bool
+    templates_dict: Optional[dict]
```
Comment (uranusjr) [inline airflow/utils/context.pyi]:
```suggestion
    templates_dict: Optional[Mapping[str, Any]]
```

File: airflow/macros/__init__.py (hunk 4 above)
Comment (potiuk) [reply airflow/macros/__init__.py]: Ah yeah. I will remove macros changes then

File: airflow/utils/context.pyi (hunk 6 above)
Comment (potiuk) [reply airflow/utils/context.pyi]: yep.

File: airflow/operators/python.py (hunk 5 above)
Comment (potiuk) [reply airflow/operators/python.py]: Ah yeah. I like it more :)

File: airflow/jobs/scheduler_job.py (hunk 3 above)
Comment (potiuk) [reply airflow/jobs/scheduler_job.py]: I feel that  since we have "open" stated explicitly, this is really "coincidentally" matching some of the task states  - especially that we only have a small subsets of the states reflected here. I'd rather create a separate enum for that.

Comment (potiuk) [reply airflow/jobs/scheduler_job.py]: Added a separate enum for that .

File: airflow/example_dags/example_task_group_decorator.py (hunk 1 above)
Comment (potiuk) [reply airflow/example_dags/example_task_group_decorator.py]: Cool

**Comment by potiuk** [comment]
All addressed !

File: airflow/example_dags/example_task_group_decorator.py (hunk 1 above)
Comment (ashb) [reply airflow/example_dags/example_task_group_decorator.py]: It does, but generally in the examples I think we should use it without parentheses

File: chart/templates/_helpers.yaml [hunk 7]
```diff
@@ -37,75 +37,56 @@ If release name contains chart name it will be used as a full name.
 {{/* Standard Airflow environment variables */}}
 {{- define "standard_airflow_environment" }}
   # Hard Coded Airflow Envs
-  {{- if .Values.enableBuiltInSecretEnvVars.AIRFLOW__CORE__FERNET_KEY }}
```
Comment (ashb) [inline chart/templates/_helpers.yaml]: This seems unrelated to mypy?

File: chart/values.schema.json [hunk 8]
```diff
@@ -725,59 +725,6 @@
                 }
             ]
         },
-        "enableBuiltInSecretEnvVars": {
-            "description": "Uses built-in secret values set as environment variables passed to Airflow. You should supply corresponding environment variables as ``extraEnv`` variables if you disable them here.",
```
Comment (kaxil) [inline chart/values.schema.json]: This change looks like a mistake

**Review by kaxil — CHANGES_REQUESTED** [review]

File: airflow/example_dags/example_task_group_decorator.py (hunk 1 above)
Comment (potiuk) [reply airflow/example_dags/example_task_group_decorator.py]: MyPy complans here if I do not use parentheses. I will double check though

File: chart/templates/_helpers.yaml (hunk 7 above)
Comment (potiuk) [reply chart/templates/_helpers.yaml]: Ups. bad rebase it seems

File: airflow/example_dags/example_task_group_decorator.py (hunk 1 above)
Comment (potiuk) [reply airflow/example_dags/example_task_group_decorator.py]:
```
airflow/example_dags/example_task_group_decorator.py:73: error: Value of type
variable "T" of "task_group_function" cannot be "int"
            current_task_group = task_group_function(i)
```

File: chart/templates/_helpers.yaml (hunk 7 above)
Comment (potiuk) [reply chart/templates/_helpers.yaml]: removed

File: chart/values.schema.json (hunk 8 above)
Comment (potiuk) [reply chart/values.schema.json]: yeah. Bad rebase.

Comment (potiuk) [reply chart/values.schema.json]: Fixed.

File: airflow/example_dags/example_task_group_decorator.py (hunk 1 above)
Comment (uranusjr) [reply airflow/example_dags/example_task_group_decorator.py]: I think `task_group`’s `if callable(python_callable)` case needs a `cast`.

Comment (uranusjr) [reply airflow/example_dags/example_task_group_decorator.py]: Oh it’s more complicated than that. I’ll just push a solution to your branch 🙂

Comment (uranusjr) [reply airflow/example_dags/example_task_group_decorator.py]: Done

File: airflow/decorators/task_group.py [hunk 9]
```diff
@@ -21,28 +21,60 @@
...
+if TYPE_CHECKING:
+    from airflow.models import DAG
+
+F = TypeVar("F", bound=Callable[..., Any])
 
 task_group_sig = signature(TaskGroup.__init__)
 
 
-def task_group(python_callable: Optional[Callable] = None, *tg_args, **tg_kwargs) -> Callable[[T], T]:
+# This covers the @task_group() case. Annotations are copied from the TaskGroup
```
Comment (potiuk) [inline airflow/decorators/task_group.py:36]: Nice. I was sure I saw the recipe somewhere :) https://mypy.readthedocs.io/en/stable/generics.html#declaring-decorators

Thanks!. The Type Hiting of Python is an interesting one.

**Comment by potiuk** [comment]
@kaxil @uranusjr @ashb - good to go?

**Review by kaxil — APPROVED** [review]
